indexer/osm_editor.cpp

```diff
@@ -0,0 +1,622 @@
+  ftInfo.m_status = FeatureStatus::Deleted;
+  ftInfo.m_feature = feature;
+  // TODO: What if local client time is absolutely wrong?
+  ftInfo.m_modificationTimestamp = time(nullptr);
+
+  // TODO(AlexZ): Synchronize Save call/make it on a separate thread.
+  Save(GetEditorFilePath());
+
+  if (m_invalidateFn)
+    m_invalidateFn();
+}
+
+//namespace
+//{
+//FeatureID GenerateNewFeatureId(FeatureID const & oldFeatureId)
+//{
+//  // TODO(AlexZ): Stable & unique features ID generation.
+//  static uint32_t newIndex = 0x0effffff;
+//  return FeatureID(oldFeatureId.m_mwmId, newIndex++);
+//}
+//}  // namespace
+
+void Editor::EditFeature(FeatureType const & editedFeature, string const & editedStreet,
+                         string const & editedHouseNumber)
+{
+  // TODO(AlexZ): Check if feature has not changed and reset status.
+  FeatureID const fid = editedFeature.GetID();
+  FeatureTypeInfo & fti = m_features[fid.m_mwmId][fid.m_index];
+  fti.m_status = FeatureStatus::Modified;
+  fti.m_feature = editedFeature;
+  // TODO: What if local client time is absolutely wrong?
+  fti.m_modificationTimestamp = time(nullptr);
+
+  fti.m_street = editedStreet;
+  if (editedHouseNumber.empty() || feature::IsHouseNumber(editedHouseNumber))
+    fti.m_feature.SetHouseNumber(editedHouseNumber);
+  // TODO(AlexZ): Store edited house number as house name if feature::IsHouseNumber() returned false.
+
+  // TODO(AlexZ): Synchronize Save call/make it on a separate thread.
+  Save(GetEditorFilePath());
+
+  if (m_invalidateFn)
+    m_invalidateFn();
+}
+
+void Editor::ForEachFeatureInMwmRectAndScale(MwmSet::MwmId const & id,
+                                             TFeatureIDFunctor const & f,
+                                             m2::RectD const & rect,
+                                             uint32_t /*scale*/)
+{
+  auto const mwmFound = m_features.find(id);
+  if (mwmFound == m_features.end())
+    return;
+
+  // TODO(AlexZ): Check that features are visible at this scale.
+  // Process only new (created) features.
+  for (auto const & index : mwmFound->second)
+  {
+    FeatureTypeInfo const & ftInfo = index.second;
+    if (ftInfo.m_status == FeatureStatus::Created &&
+        rect.IsPointInside(ftInfo.m_feature.GetCenter()))
+      f(FeatureID(id, index.first));
+  }
+}
+
+void Editor::ForEachFeatureInMwmRectAndScale(MwmSet::MwmId const & id,
+                                             TFeatureTypeFunctor const & f,
+                                             m2::RectD const & rect,
+                                             uint32_t /*scale*/)
+{
+  auto mwmFound = m_features.find(id);
+  if (mwmFound == m_features.end())
+    return;
+
+  // TODO(AlexZ): Check that features are visible at this scale.
+  // Process only new (created) features.
+  for (auto & index : mwmFound->second)
+  {
+    FeatureTypeInfo & ftInfo = index.second;
+    if (ftInfo.m_status == FeatureStatus::Created &&
```

Comment:
They are processed separately, by other methods. These ForEach* used to query new features which are absent in our mwm files.